### PR TITLE
Change selection of Python interpreter for devpi upload.

### LINKED ISSUE #395: Upload runs with wrong Python
- Bitbucket: https://bitbucket.org/hpk42/devpi/issue/395
- Originally reported by: @jaraco
- Originally created at: 2017-03-10T19:45:16.023

Trying to use `devpi upload --only-docs` the run crashes because it's grabbing an old Python that's not compatible with the environment. The same issue occurs with a plain `upload`:

```
$ rwt sphinx -r docs/requirements.txt -- -m devpi upload             
Collecting sphinx
  Using cached Sphinx-1.5.3-py2.py3-none-any.whl
Collecting rst.linker (from -r docs/requirements.txt (line 1))
  Using cached rst.linker-1.8.2-py2.py3-none-any.whl
Collecting Pygments>=2.0 (from sphinx)
  Using cached Pygments-2.2.0-py2.py3-none-any.whl
Collecting snowballstemmer>=1.1 (from sphinx)
  Using cached snowballstemmer-1.2.1-py2.py3-none-any.whl
Collecting requests>=2.0.0 (from sphinx)
  Using cached requests-2.13.0-py2.py3-none-any.whl
Collecting babel!=2.0,>=1.3 (from sphinx)
  Using cached Babel-2.3.4-py2.py3-none-any.whl
Collecting Jinja2>=2.3 (from sphinx)
  Using cached Jinja2-2.9.5-py2.py3-none-any.whl
Collecting six>=1.5 (from sphinx)
  Using cached six-1.10.0-py2.py3-none-any.whl
Collecting alabaster<0.8,>=0.7 (from sphinx)
  Using cached alabaster-0.7.10-py2.py3-none-any.whl
Collecting docutils>=0.11 (from sphinx)
  Using cached docutils-0.13.1-py3-none-any.whl
Collecting imagesize (from sphinx)
  Using cached imagesize-0.7.1-py2.py3-none-any.whl
Collecting python-dateutil (from rst.linker->-r docs/requirements.txt (line 1))
  Using cached python_dateutil-2.6.0-py2.py3-none-any.whl
Collecting pytz>=0a (from babel!=2.0,>=1.3->sphinx)
  Using cached pytz-2016.10-py2.py3-none-any.whl
Collecting MarkupSafe>=0.23 (from Jinja2>=2.3->sphinx)
  Using cached MarkupSafe-1.0.tar.gz
Installing collected packages: Pygments, snowballstemmer, requests, pytz, babel, MarkupSafe, Jinja2, six, alabaster, docutils, imagesize, sphinx, python-dateutil, rst.linker
  Running setup.py install for MarkupSafe ... done
Successfully installed Jinja2-2.9.5 MarkupSafe-1.0 Pygments-2.2.0 alabaster-0.7.10 babel-2.3.4 docutils-0.13.1 imagesize-0.7.1 python-dateutil-2.6.0 pytz-2016.10 requests-2.13.0 rst.linker-1.8.2 six-1.10.0 snowballstemmer-1.2.1 sphinx-1.5.3
detected devpi:upload section in /Users/jaraco/Dropbox/code/yg/G/queso/setup.cfg
using workdir /var/folders/c6/v7hnmq453xb6p2dbz1gqc6rr0000gn/T/devpi2
copied repo /Users/jaraco/Dropbox/code/yg/G/queso/.git to /var/folders/c6/v7hnmq453xb6p2dbz1gqc6rr0000gn/T/devpi2/upload/queso/.git
pre-build: cleaning /Users/jaraco/Dropbox/code/yg/G/queso/dist
-->  /var/folders/c6/v7hnmq453xb6p2dbz1gqc6rr0000gn/T/devpi2/upload/queso$ /usr/local/bin/python setup.py sdist --formats gztar 
Traceback (most recent call last):
  File "setup.py", line 75, in <module>
    setuptools.setup(**params)
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/distutils/core.py", line 151, in setup
    dist.run_commands()
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/distutils/dist.py", line 953, in run_commands
    self.run_command(cmd)
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/distutils/dist.py", line 972, in run_command
    cmd_obj.run()
  File "/usr/local/lib/python2.7/site-packages/setuptools/command/sdist.py", line 51, in run
    self.run_command(cmd_name)
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/distutils/cmd.py", line 326, in run_command
    self.distribution.run_command(command)
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/distutils/dist.py", line 970, in run_command
    cmd_obj = self.get_command_obj(command)
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/distutils/dist.py", line 845, in get_command_obj
    klass = self.get_command_class(command)
  File "/usr/local/lib/python2.7/site-packages/setuptools/dist.py", line 492, in get_command_class
    return _Distribution.get_command_class(self, command)
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/distutils/dist.py", line 815, in get_command_class
    __import__ (module_name)
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/distutils/command/check.py", line 13, in <module>
    from docutils.utils import Reporter
  File "/var/folders/c6/v7hnmq453xb6p2dbz1gqc6rr0000gn/T/rwt-igrsglo7/docutils/utils/__init__.py", line 20, in <module>
    import docutils.io
  File "/var/folders/c6/v7hnmq453xb6p2dbz1gqc6rr0000gn/T/rwt-igrsglo7/docutils/io.py", line 340
    (self.destination.mode, mode)), file=self._stderr)
                                        ^
SyntaxError: invalid syntax
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/devpi/__main__.py", line 3, in <module>
    main.main()
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/devpi/main.py", line 32, in main
    return method(hub, hub.args)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/devpi/upload.py", line 41, in main
    default_formats=setupcfg.get("formats")))
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/devpi/upload.py", line 401, in setup_build
    out = self.hub.popen_output(cmd, cwd=self.rootpath)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/devpi/main.py", line 235, in popen_output
    return check_output(args, cwd=str(cwd))
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/devpi_common/proc.py", line 18, in check_output
    raise CalledProcessError(retcode, cmd, output=output)
subprocess.CalledProcessError: Command '['/usr/local/bin/python', 'setup.py', 'sdist', '--formats', 'gztar']' returned non-zero exit status 1.
```

The program fails because it's invoking the system python (/usr/local/bin/python), which resolves to Python2.7, which isn't compatible with the Python 3 libraries installed for sphinx.

Obviously, there I'm using rwt to resolve the sphinx and doc build dependencies, but you can encounter the same issue with a virtualenv:

```
$ python -m venv ~/.envs/devpi-doc-upload
$ ~/.envs/devpi-doc-upload/bin/pip install sphinx -r docs/requirements.txt 
Collecting sphinx
  Using cached Sphinx-1.5.3-py2.py3-none-any.whl
Collecting rst.linker (from -r docs/requirements.txt (line 1))
  Using cached rst.linker-1.8.2-py2.py3-none-any.whl
Collecting imagesize (from sphinx)
  Using cached imagesize-0.7.1-py2.py3-none-any.whl
Collecting Jinja2>=2.3 (from sphinx)
  Using cached Jinja2-2.9.5-py2.py3-none-any.whl
Collecting alabaster<0.8,>=0.7 (from sphinx)
  Using cached alabaster-0.7.10-py2.py3-none-any.whl
Collecting requests>=2.0.0 (from sphinx)
  Using cached requests-2.13.0-py2.py3-none-any.whl
Collecting six>=1.5 (from sphinx)
  Using cached six-1.10.0-py2.py3-none-any.whl
Collecting docutils>=0.11 (from sphinx)
  Using cached docutils-0.13.1-py3-none-any.whl
Collecting babel!=2.0,>=1.3 (from sphinx)
  Using cached Babel-2.3.4-py2.py3-none-any.whl
Collecting Pygments>=2.0 (from sphinx)
  Using cached Pygments-2.2.0-py2.py3-none-any.whl
Collecting snowballstemmer>=1.1 (from sphinx)
  Using cached snowballstemmer-1.2.1-py2.py3-none-any.whl
Collecting python-dateutil (from rst.linker->-r docs/requirements.txt (line 1))
  Using cached python_dateutil-2.6.0-py2.py3-none-any.whl
Collecting MarkupSafe>=0.23 (from Jinja2>=2.3->sphinx)
  Using cached MarkupSafe-1.0.tar.gz
Collecting pytz>=0a (from babel!=2.0,>=1.3->sphinx)
  Using cached pytz-2016.10-py2.py3-none-any.whl
Installing collected packages: imagesize, MarkupSafe, Jinja2, alabaster, requests, six, docutils, pytz, babel, Pygments, snowballstemmer, sphinx, python-dateutil, rst.linker
  Running setup.py install for MarkupSafe ... done
Successfully installed Jinja2-2.9.5 MarkupSafe-1.0 Pygments-2.2.0 alabaster-0.7.10 babel-2.3.4 docutils-0.13.1 imagesize-0.7.1 python-dateutil-2.6.0 pytz-2016.10 requests-2.13.0 rst.linker-1.8.2 six-1.10.0 snowballstemmer-1.2.1 sphinx-1.5.3
$ ~/.envs/devpi-doc-upload/bin/pip install devpi-client
Collecting devpi-client
  Downloading devpi_client-2.7.0-py2.py3-none-any.whl
Collecting devpi-common<4.0,>2.0.2 (from devpi-client)
  Downloading devpi_common-3.0.1-py2.py3-none-any.whl
Collecting tox>=1.7.1 (from devpi-client)
  Using cached tox-2.6.0-py2.py3-none-any.whl
Collecting pkginfo>=1.2b1 (from devpi-client)
  Downloading pkginfo-1.4.1-py2.py3-none-any.whl
Collecting py>=1.4.31 (from devpi-client)
  Using cached py-1.4.32-py2.py3-none-any.whl
Collecting check-manifest>=0.28 (from devpi-client)
  Cache entry deserialization failed, entry ignored
  Using cached check_manifest-0.35-py2.py3-none-any.whl
Requirement already satisfied: requests>=2.3.0 in /Users/jaraco/.envs/devpi-doc-upload/lib/python3.6/site-packages (from devpi-common<4.0,>2.0.2->devpi-client)
Collecting pluggy<1.0,>=0.3.0 (from tox>=1.7.1->devpi-client)
  Downloading pluggy-0.4.0-py2.py3-none-any.whl
Collecting virtualenv>=1.11.2; python_version != "3.2" (from tox>=1.7.1->devpi-client)
  Downloading virtualenv-15.1.0-py2.py3-none-any.whl (1.8MB)
    100% |████████████████████████████████| 1.8MB 601kB/s 
Installing collected packages: py, devpi-common, pluggy, virtualenv, tox, pkginfo, check-manifest, devpi-client
Successfully installed check-manifest-0.35 devpi-client-2.7.0 devpi-common-3.0.1 pkginfo-1.4.1 pluggy-0.4.0 py-1.4.32 tox-2.6.0 virtualenv-15.1.0



$ ~/.envs/devpi-doc-upload/bin/python -m devpi upload --only-docs
detected devpi:upload section in /Users/jaraco/Dropbox/code/yg/G/queso/setup.cfg
using workdir /var/folders/c6/v7hnmq453xb6p2dbz1gqc6rr0000gn/T/devpi4
copied repo /Users/jaraco/Dropbox/code/yg/G/queso/.git to /var/folders/c6/v7hnmq453xb6p2dbz1gqc6rr0000gn/T/devpi4/upload/queso/.git
pre-build: cleaning /Users/jaraco/Dropbox/code/yg/G/queso/dist
-->  /var/folders/c6/v7hnmq453xb6p2dbz1gqc6rr0000gn/T/devpi4/upload/queso$ /usr/local/bin/python setup.py build_sphinx -E --build-dir /var/folders/c6/v7hnmq453xb6p2dbz1gqc6rr0000gn/T/devpi4/upload/queso/build 
usage: setup.py [global_opts] cmd1 [cmd1_opts] [cmd2 [cmd2_opts] ...]
   or: setup.py --help [cmd1 cmd2 ...]
   or: setup.py --help-commands
   or: setup.py cmd --help

error: invalid command 'build_sphinx'
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/Users/jaraco/.envs/devpi-doc-upload/lib/python3.6/site-packages/devpi/__main__.py", line 3, in <module>
    main.main()
  File "/Users/jaraco/.envs/devpi-doc-upload/lib/python3.6/site-packages/devpi/main.py", line 32, in main
    return method(hub, hub.args)
  File "/Users/jaraco/.envs/devpi-doc-upload/lib/python3.6/site-packages/devpi/upload.py", line 43, in main
    p = exported.setup_build_docs()
  File "/Users/jaraco/.envs/devpi-doc-upload/lib/python3.6/site-packages/devpi/upload.py", line 417, in setup_build_docs
    "--build-dir", build], cwd=self.rootpath)
  File "/Users/jaraco/.envs/devpi-doc-upload/lib/python3.6/site-packages/devpi/main.py", line 219, in popen_output
    return check_output(args, cwd=str(cwd))
  File "/Users/jaraco/.envs/devpi-doc-upload/lib/python3.6/site-packages/devpi_common/proc.py", line 18, in check_output
    raise CalledProcessError(retcode, cmd, output=output)
subprocess.CalledProcessError: Command '['/usr/local/bin/python', 'setup.py', 'build_sphinx', '-E', '--build-dir', '/var/folders/c6/v7hnmq453xb6p2dbz1gqc6rr0000gn/T/devpi4/upload/queso/build']' returned non-zero exit status 1.
```

The reason `build_sphinx` is missing is because the Python being used is the system Python 2.7 which doesn't have Sphinx installed.

I haven't yet looked into the code, but I suspect that these build steps should use `sys.executable` instead of `'python'` for their subprocess operations.

### PR DESCRIPTION
**Pull request by fschulze**
If used the new ``-p/--python`` option has priority, then a currently activated virtualenv, lastly the ``sys.executable`` under which devpi is running.

Fixes #395 

If possible, please try this change @jaraco @wimglenn @cantzler @pawciobiel @smsearcy @r2dan 
I've tried the scenarios I could think of and it seems to do the right thing.

A package is available here:
https://m.devpi.net/devpi-travis/fschulze_devpi_395-upload-python-selection/devpi-client/5.0.0.dev0

### CODE REVIEW

File: client/devpi/upload.py [hunk 1]
```diff
@@ -334,10 +334,25 @@ def __init__(self, hub, rootpath, origrepo):
...
+
+    @property
+    def python(self):
+        """
+        Find the best Python executable for invoking builds and uploads.
+        Command line option if used has priority, then a virtualenv if one
+        is activated, but otherwise falls back to sys.executable, the Python
+        under which devpi client is running.
+        """
+        python = self.hub.args.python
```
Comment (jaraco) [inline client/devpi/upload.py]: I would have written this as:

```
return self.hub.args.python or self._virtualenv_python() or sys.executable
```

**Comment by jaraco** [comment]
Looks good to me. I realize we had worked around the issue by relying on `tox` to run our releases, in which case, our devpi-client is installed in a virtualenv and everything works on the existing code.

Still, I think this behavior is an improvement in general.

File: client/devpi/upload.py (hunk 1 above)
Comment (fschulze) [reply client/devpi/upload.py]: Yes, you had that in your original diff. I think it's harder to read though, that's why I split it up.

**Comment by AWhetter** [comment]
Are there any plans to get this fixed up? If not I'd be happy to get this finished off.

**Comment by HarrySky** [comment]
Is there any plans for releasing 5.0.0 soon? This feature is needed

**Comment by fschulze** [comment]
I plan to do releases by the end of this week.

**Comment by fschulze** [comment]
I released devpi-client 5.0.0

**Pull request merged**